add quantized groupnorm operator

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37802 quant: clean up normalization channels_last handling
* #37781 quant: add test util for testing off by one quantized tensors
* #37352 move quantization normalization layers to aten/src/ATen/native/quantized/cpu/
* #36847 add quantized instancenorm operator
* **#36835 add quantized groupnorm operator**
* #37947 quant: remove hypothesis and int32 from layernorm test

Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

vkuzo committed:
add quantized groupnorm operator

Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -165,7 +165,60 @@ Tensor quantized_layer_norm_impl(
+  return Y;
+}
+
+Tensor quantized_group_norm_impl(
```
Is there a reason why this isn't defined in the native/quantized/cpu folder?

vkuzo replied:
It was mostly convenience of reusing `_prepare_layer_norm_inputs` which is used by LayerNorm.  Would `native/quantized/cpu` be a better place for this / is it worth refactoring? I'm flexible

supriyar replied:
I think it would be good to have all the quantized ops live in the same folder. You should be able to call `_prepare_layer_norm_inputs` from there. i think we do something similar in qpool.cpp where we reuse the output shape computation logic.

vkuzo replied:
fixed this in https://github.com/pytorch/pytorch/pull/36835, hope that's ok!

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -141,6 +141,32 @@ def _test_hardsigmoid(self, X, engine):
+        num_bins = 2 ** 32
+        X_scale = float(max_val - min_val) / num_bins
+    elif torch_type == torch.qint8:
+        X_zero_point = 0
```
Can we test with a non-zero zero-point?

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -1565,6 +1571,92 @@ def test_batch_norm2d(self, X, Y_scale, Y_zero_point):
+
+            # In the FP kernel, mean and variance are calculated in floating point.
+            # In the quantized kernel, they are calculated in integer arithmetic.
+            # Because of this, the numerics do not always match exactly which is
```
Can we also point out that the quantized computation is exact, as we do not incur any precision loss.

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized groupnorm operator"


Summary:

Adds a quantized groupnorm operator.  We reuse most of the layernorm
kernel, modifying it to be able to perform channel-wise scaling.

Benchmark results: the quantized layer is between 6x to 15x faster
from fp to q, depending on input shapes
(full results:
https://gist.github.com/vkuzo/db67623232415382dabff6c8923124e9)

Test Plan:

```
python test/quantization/test_quantized.py TestQuantizedOps.test_group_norm
python test/quantization/test_quantized.py TestQuantizedOps.test_qlayer_norm
```

Numerics are nearly equivalent, with the only difference documented
in the test case.  The difference is the same type as with quantized
layernorm.  Making numerics equivalent is possible but will sacrifice
speed.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21107926](https://our.internmc.facebook.com/intern/diff/D21107926)

[ghstack-poisoned]
